### PR TITLE
fix(batch): fix dangling channel and prevent receiver from infinite waiting

### LINKED ISSUE #5816: bug(batch): cyclic reference in QueryRunner causes dangling tokio tasks
### Describe the bug

When a `QueryRunner` is created, a `msg_receiver` is passed in to notify `QueryRunner` that all stages have been scheduled when it runs. However, the `msg_sender`s owned by `StageExecution`s (which are owned by `QueryRunner`) are never dropped, causing a cyclic reference.

```
QueryRunner::run (poll the receiver) -> StageExecution (contains the senders)
                            ⬆️                                      ⬇️ 
                            -----------------------------------------
```

We were not able to identify the bug until now, because the join handle of the task spawned to call `QueryRunner::run` is ignored. So even if the query finishes, there will be a tokio task dangling around.

https://github.com/risingwavelabs/risingwave/blob/6364145c5aae2b73d9277055ab06026906c23a4f/src/frontend/src/scheduler/distributed/query.rs#L172-L184

### Possible Solution 

- In `StageExecution`, drop the `msg_sender` when needed so that the cycle can be broken.
- Also, we need to join all the tasks that have been spawned.

### To Reproduce

Run the snippet in #5724 

### Expected behavior

All the tasks related to a query should be joined when it finishes.

### Additional context

This is one of the causes of #5724

### PR DESCRIPTION
**Pull request by xx01cyx**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

### Problem

When fixing #5724 (which is caused by #5787) , we find that a receiver in `QueryRunner` waits infinitely in a `while` loop. To be more specific, it waits for messages coming from the senders, while the senders have finished sending all messages, but fail to be dropped because of some cyclic reference issues. View #5816 for more details.

### Solution

Align lifetime of the sender to that of `StageRunner` instead of `StageExecution`, so it will be dropped immediately when `StageRunner` ends its job. When all senders are dropped, the channel will be closed then, and the receiver will be dropped accordingly.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)

Fixes #5816

### CODE REVIEW

File: src/frontend/src/scheduler/distributed/stage.rs [hunk 1]
```diff
@@ -183,14 +181,15 @@ impl StageExecution {
     /// Starts execution of this stage, returns error if already started.
     pub async fn start(&self) {
         let mut s = self.state.write().await;
-        match &*s {
-            &StageState::Pending => {
+        let cur_state = mem::replace(&mut *s, StageState::Failed);
```
Comment (Gun9niR) [inline src/frontend/src/scheduler/distributed/stage.rs:184]: Why should we replace the state with `Failed`? Even though it won't cause any bugs, it looks counter-intuitive. Same as 

https://github.com/risingwavelabs/risingwave/blob/b5f84f32e0f20b59f161bef803bbb1a4dfd51b8a/src/frontend/src/scheduler/distributed/query.rs#L163

Here, `Failed` is just a dummy value.

File: src/frontend/src/scheduler/distributed/stage.rs [hunk 2]
```diff
@@ -183,14 +181,15 @@ impl StageExecution {
...
+        match cur_state {
+            StageState::Pending { msg_sender } => {
```
Comment (Gun9niR) [inline src/frontend/src/scheduler/distributed/stage.rs:186]: Should we add some comments here to explain why we put the sender in `state` (or just add a note above `QueryExecution::shutdown_tx`? This cyclic reference is pretty tricky, and future developers should be warned of it.

**Review by Gun9niR — APPROVED** [review]
Rest LGTM.

File: src/frontend/src/scheduler/distributed/stage.rs (hunk 1 above)
Comment (xx01cyx) [reply src/frontend/src/scheduler/distributed/stage.rs:184]: cc. @liurenjie1024

**Comment by BugenZhao** [comment]
Can we also store the `JoinHandle` in the query result fetcher, and `join` it when the stream is done? So that we can never worry about this issue in the future. 🤪

File: src/frontend/src/scheduler/distributed/stage.rs (hunk 1 above)
Comment (BowenXiao1999) [reply src/frontend/src/scheduler/distributed/stage.rs:190]: I think this is try to avoid rust compiler complain on lifetime and ownership?

**Review by BowenXiao1999 — COMMENTED** [review]
LGTM

**Comment by xx01cyx** [comment]
> Can we also store the `JoinHandle` in the query result fetcher, and join it when the stream is done? So that we can never worry about this issue in the future.

Currently we have not cleaned `QueryExecution` when the query ends (this will be done in my future PR). Therefore, the `shutdown_tx` in `QueryExecution` will not be dropped as well. As a result, the channel cannot get closed, and the receiver will still get stuck in that [loop](https://github.com/risingwavelabs/risingwave/blob/main/src/frontend/src/scheduler/distributed/query.rs#L233-L289). So the spawned threads are not able to get joined currently. I'll get this done in my future PR, after I clean up all the  `QueryExecution`s.

**Review by BowenXiao1999 — APPROVED** [review]

File: src/frontend/src/scheduler/distributed/stage.rs (hunk 1 above)
Comment (BowenXiao1999) [reply src/frontend/src/scheduler/distributed/stage.rs:190]: Anyway, feel free to submit PR to fix this if you have better impl with the same semantics 😆

**Review by BugenZhao — APPROVED** [review]

**Pull request merged**